static/app/components/charts/useChartXRangeSelection.tsx

```diff
@@ -0,0 +1,415 @@
+   */
+  chartsGroupName?: string;
+
+  /**
+   * The dependencies to be used to re-activate selection or re-paint the box.
+   */
+  deps?: DependencyList;
+
+  /**
+   * Whether selection is disabled.
+   */
+  disabled?: boolean;
+
+  /**
+   * The callback that is called when the selection is cleared.
+   */
+  onClearSelection?: () => void;
+
+  /**
+   * The callback that is called when the selection/dragging ends.
+   */
+  onSelectionEnd?: (selection: Selection, clearSelection: () => void) => void;
+
+  /**
+   * The callback that is called when the selection/dragging starts.
+   */
+  onSelectionStart?: () => void;
+};
+
+export function useChartXRangeSelection({
+  chartRef,
+  onSelectionEnd,
+  onSelectionStart,
+  onClearSelection,
+  actionMenuRenderer,
+  chartsGroupName,
+  disabled = false,
+  deps = [],
+}: ChartXRangeSelectionProps): BoxSelectionOptions {
+  const [state, setState] = useState<State>();
+  const tooltipFrameRef = useRef<number | null>(null);
+  const enabbleBrushModeFrameRef = useRef<number | null>(null);
+
+  const onBrushStart = useCallback<EChartBrushStartHandler>(
+    (_evt: any, chartInstance: any) => {
+      // Echarts either lets us connect all interactivity of the charts in a group or none of them.
+      // We need connectivity for cursor syncing, but having that enabled while drawing, leads to a
+      // box drawn for all of the charts in the group. We are going for chart specific box selections,
+      // so we disconnect the group while drawing.
+      if (chartsGroupName) {
+        echarts?.disconnect(chartsGroupName);
+      }
+
+      chartInstance.dispatchAction({type: 'hideTip'});
+
+      // Disable the tooltip as we start dragging, as it covers regions of the chart that the user
+      // may want to select. The tooltip remains hidden until the box is cleared.
+      if (tooltipFrameRef.current) cancelAnimationFrame(tooltipFrameRef.current);
+
+      tooltipFrameRef.current = requestAnimationFrame(() => {
+        chartInstance.setOption(
+          {
+            tooltip: {show: false},
+          },
+          {silent: true}
+        );
+      });
+
+      onSelectionStart?.();
+    },
+    [chartsGroupName, onSelectionStart]
+  );
+
+  const clearSelection = useCallback(() => {
+    if (!state?.selection) return;
+
+    const chartInstance = chartRef.current?.getEchartsInstance();
+
+    chartInstance?.dispatchAction({type: 'brush', areas: []});
+
+    // Restore the tooltip as we clear selection
+    if (tooltipFrameRef.current) cancelAnimationFrame(tooltipFrameRef.current);
+
+    tooltipFrameRef.current = requestAnimationFrame(() => {
+      chartInstance?.setOption({tooltip: {show: true}}, {silent: true});
+    });
+
+    setState(null);
+
+    onClearSelection?.();
+  }, [chartRef, onClearSelection, state?.selection]);
+
+  const onBrushEnd = useCallback<EChartBrushEndHandler>(
+    (evt: any, chartInstance: any) => {
```

Comment:
Same question about types as `onBrushStart`

Reply:
Addressed in the comment above.